### ISSUE TITLE
Get link to API

### ISSUE THREAD

**Issue by Arka21Pal** [Arka21Pal/aws-practice#12]
Right now, because AWS creates an unique ID every time I create an API, I can't statically link the API to my website.

There are 2 solutions here:

1. I buy a domain name and link my API to that domain name every time. This is a robust solution.
2. I get the new domain name of the API every time I deploy the template and insert it into my website.

**Comment by Arka21Pal** [comment]
AWS cli command to get link to API:
```
aws apigatewayv2 get-apis --query "Items[].ApiEndpoint | [-1]" --profile <profile> --region <region>
```

References:

1. https://docs.aws.amazon.com/cli/latest/reference/apigatewayv2/get-apis.html
2. https://jmespath.org/tutorial.html#pipe-expressions

---

Perhaps the more convenient method would be to get the API link in `boto3` and use another lambda function to insert (and overwrite the previous link, if any) directly in my webpage.

**Comment by Arka21Pal** [comment]
A decent idea would be to store the link in a variable in a separate JS file and then access it from the main JS code by importing the file

**Comment by Arka21Pal** [comment]
- Get all APIs using `boto3`: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/apigatewayv2.html#ApiGatewayV2.Client.get_apis
    - Corresponding AWS documentation: https://docs.aws.amazon.com/apigatewayv2/latest/api-reference/apis.html#apisget
- The idea will be to get the API ID from the command specified above, then to use a lambda function to write to a JS/JSON file with the variable and the link for the API. This URL will then be passed into the main JS function for the website.

Both the Lambda Function and the bucket into which the Lambda function will write will need to go into  the `support-stack`.

**Comment by Arka21Pal** [comment]
Invoke Lambda function from code: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/lambda.html#Lambda.Client.invoke

**Comment by Arka21Pal** [comment]
- S3 notification configuration (trigger): https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket-notificationconfig.html
    - `AWS::S3::Bucket LambdaConfiguration` (for notification): https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket-notificationconfig-lambdaconfig.html
        - Event configuration: https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-how-to-event-types-and-destinations.html
        - `AWS::S3::Bucket NotificationFilter`: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket-notificationconfiguration-config-filter.html (basically, just specify the filetype or the file/S3key)
            - `AWS::S3::Bucket S3KeyFilter`: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket-notificationconfiguration-config-filter-s3key.html
                - `AWS::S3::Bucket FilterRule`: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket-notificationconfiguration-config-filter-s3key-rules.html
    - Example: https://stackoverflow.com/a/46854663
- Example: https://www.itonaut.com/2018/10/03/implement-s3-bucket-lambda-triggers-in-aws-cloudformation/

**Comment by Arka21Pal** [comment]
`AWS::Lambda::Permission`: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-permission.html

> The `AWS::Lambda::Permission` resource grants an AWS service or another account permission to use a function. You can apply the policy at the function level, or specify a qualifier to restrict access to a single version or alias. If you use a qualifier, the invoker must use the full Amazon Resource Name (ARN) of that version or alias to invoke the function.

**Comment by Arka21Pal** [comment]
It would seem that I will have to first download the file to the `/tmp` directory in Lambda to be able to edit the file.

Reference: https://medium.com/@haldis444/use-lambda-to-append-daily-data-to-csv-file-in-s3-2c2813bc33d0

**Comment by Arka21Pal** [comment]
Fixed bug, completed in commit 64193001cb9ad8186b7ad4a85edb834ff896a901
